daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Normalization.scala

```diff
@@ -0,0 +1,153 @@
+  private type Node = GenNode[Nid, Cid]
+  private type VTX = VersionedTransaction[Nid, Cid]
+
+  def normalizeTx(vtx: VTX): VTX = {
+    vtx match {
+      case VersionedTransaction(_, nodes, roots) =>
+        // TODO: Normalized version calc should be shared with code in asVersionedTransaction
+        val version = roots.iterator.foldLeft(TransactionVersion.minVersion) { (acc, nodeId) =>
+          import scala.Ordering.Implicits.infixOrderingOps
+          nodes(nodeId).optVersion match {
+            case Some(version) => acc max version
+            case None => acc max TransactionVersion.minExceptions
+          }
+        }
+        VersionedTransaction(
+          version,
+          nodes.map { case (k, v) =>
+            (k, normNode(v))
+          },
+          vtx.roots,
+        )
+    }
+  }
+
+  private def normNode(
+      node: Node
+  ): Node = {
+    import scala.Ordering.Implicits.infixOrderingOps
+    node match {
+
+      case old: NodeCreate[_] =>
+        old
+          .copy(arg = normValue(old.arg))
+          .copy(key = old.key.map(normKWM))
+
+      case old: NodeFetch[_] =>
+        (if (old.version >= TransactionVersion.minByKey) {
+           old
+         } else {
+           old.copy(byKey = false)
+         })
+          .copy(
+            key = old.key.map(normKWM)
+          )
+
+      case old: NodeExercises[_, _] =>
+        (if (old.version >= TransactionVersion.minByKey) {
+           old
+         } else {
+           old.copy(byKey = false)
+         })
+          .copy(
+            chosenValue = normValue(old.chosenValue),
+            exerciseResult = old.exerciseResult.map(normValue),
+            key = old.key.map(normKWM),
+          )
+
+      case old: NodeLookupByKey[_] =>
+        old.copy(
+          key = normKWM(old.key)
+        )
+
+      case old: NodeRollback[_] => old
+
+    }
+  }
+
+  // It's ok to code this in stack unaware fashion, as values  have a depth limit of 100
+  private def normValue(x: Val): Val = {
+    x match {
+
+      // recursive cases (with normalization)
+      case V.ValueRecord(_, fields) =>
+        V.ValueRecord(
+          tycon = None, //norm
+          fields = fields.map { case (_, v) =>
+            (
+              None, //norm
+              normValue(v),
+            )
+          },
+        )
+
+      case V.ValueVariant(_, variant, v) =>
+        V.ValueVariant(
+          None, //norm
+          variant,
+          normValue(v),
+        )
+
+      // other recursive cases
+      case V.ValueList(list) => V.ValueList(list.map(normValue))
+      case V.ValueOptional(opt) => V.ValueOptional(opt.map(normValue))
+      case V.ValueGenMap(entries) =>
+        V.ValueGenMap(entries.map { case (k, v) => (normValue(k), normValue(v)) })
+      case V.ValueTextMap(x) => V.ValueTextMap(x.mapValue(normValue))
+
+      // non-recursive cases
+      case V.ValueContractId(_) => x
+      case _: V.ValueCidlessLeaf => x
```

Comment:
This isn’t correct. you need to normalise enums to drop the constructor in LF >= 1.12. I recommend to match on the individual types to avoid issues like this in the future. This is also another argument to not normalise the serialized version outside of testst which should have caught this bug.

Reply:
fixed
but then switched to existing code anyway!